### PR TITLE
Append Char to String

### PR DESCRIPTION
**Pull request by cchantep**
It would be nice to have simpler way to append a `Char` to a `String`.

```idris
||| Adds a character to the end of the specified string.
|||
||| ```idris example
||| strAppend 'A' 'B'
||| ```
||| ```idris example
||| strAppend "" 'A'
||| ```
strAppend : String -> Char -> String
strAppend s c = s ++ (strCons 'c' "")

||| Alias of strAppend
(+) : String -> Char -> String
(+) = strAppend
```

### CODE REVIEW

File: libs/prelude/Prelude/Strings.idr [hunk 1]
```diff
@@ -132,6 +132,25 @@ pack = foldr strCons ""
...
+||| ```
+strAppend : String -> Char -> String
+strAppend s c = s ++ (singleton 'c')
+
+||| Alias of `strAppend`
+|||
+||| ```idris example
+||| "AB" +> 'C'
+||| ```
+(+) : String -> Char -> String
```
Comment (cchantep) [inline libs/prelude/Prelude/Strings.idr]: Try to declare with `(+>)` I got the following error.

```
./Prelude/Strings.idr:152:1: error: Missing
    fixity declaration for
    Prelude.Strings.&, expected: "->",
    ";",
    ambiguous use of a left-associative operator,
    ambiguous use of a non-associative operator,
    ambiguous use of a right-associative operator,
    matching application expression
```

Comment (jfdm) [inline libs/prelude/Prelude/Strings.idr]: You will probably have to declare the [fixity of the operator](http://idris.readthedocs.org/en/latest/tutorial/typesfuns.html?highlight=infix#data-types).

**Comment by david-christiansen** [comment]
I'd like to see the following prior to a merge:

 * Put these into the contrib/ package, in Data.String (creating the module if necessary). New things should go there, unless they're trivial proofs about existing things in the Prelude or unless there's some other major reason to put it in the Prelude.
 * Have both `<+` and `+>` operators, the former as a synonym for `strCons`
 * Make `+>` be left-associative, so that one can write "aubergi" +> 'n' +> 'e'
 * Rename `strAppend` to `strSnoc` instead, for more consistency with other names

Thanks for the contribution!

**Comment by cchantep** [comment]
Updated

File: libs/contrib/Data/String.idr [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+(+>) : String -> Char -> String
+(+>) = strSnoc
+
+||| Alias of `strCons`
+|||
+||| ```idris example
+||| 'A' <+ "AB"
+||| ```
+(<+) : Char -> String -> String
+(<+) = strCons
```
Comment (Heather) [inline libs/contrib/Data/String.idr]: How should this work?

**Comment by cchantep** [comment]
CI issue?

**Comment by david-christiansen** [comment]
Sorry, I totally didn't notice that you had re-organized this! I just restarted Travis for the failed case and will merge when it's done.

Thanks for your patience, and I'm sorry to have dropped the ball here!

**Comment by Heather** [comment]
@david-christiansen I'm sorry but may you tell me how does `strSnoc` will work for both `String -> Char -> String` and `Char -> String -> String` in the same time? Is it some language feature that I don't know...

File: libs/contrib/Data/String.idr [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+||| Adds a character to the end of the specified string.
+|||
+||| ```idris example
+||| strSnoc "AB" 'C'
+||| ```
+||| ```idris example
+||| strSnoc "" 'A'
+||| ```
+strSnoc : String -> Char -> String
+strSnoc s c = s ++ (singleton 'c')
```
Comment (melted) [inline libs/contrib/Data/String.idr]: This appends a literal `c`...

**Comment by david-christiansen** [comment]
@Heather It won't - `strCons` is putting a char at the front, and `strSnoc` is putting it at the end. `strCons` has been in the Idris prelude for a long time. It's a wrapper around `prim__strCons` which does the same thing.

**Comment by Heather** [comment]
@david-christiansen I see, I miss-read calling function name :)

**Comment by david-christiansen** [comment]
Thanks again @cchantep!

**Comment by melted** [comment]
Nooooooo!

----- Ursprungligt meddelande -----
Från: "David Christiansen" <notifications@github.com>
Skickat: ‎2015-‎04-‎20 15:07
Till: "idris-lang/Idris-dev" <Idris-dev@noreply.github.com>
Kopia: "Niklas Larsson" <metaniklas@gmail.com>
Ämne: Re: [Idris-dev] Append Char to String (#2048)

Merged #2048.

**Comment by david-christiansen** [comment]
Thanks for the catch - I misunderstood your comment there.

**Pull request merged**